processing/src/main/java/org/apache/druid/query/operator/WindowOperatorQueryQueryToolChest.java

```diff
@@ -116,6 +128,36 @@ public Sequence<Object[]> resultsAsArrays(
     return (Sequence) resultSequence;
   }
 
+  @Override
+  public Optional<Sequence<FrameSignaturePair>> resultsAsFrames(
+      WindowOperatorQuery query,
+      Sequence<RowsAndColumns> resultSequence,
+      MemoryAllocatorFactory memoryAllocatorFactory,
+      boolean useNestedForUnknownTypes
+  )
+  {
```

Comment:
Generally speaking, this methods implementation is entirely incorrect.  RACs already know how to turn themselves into frames, this should just be doing a map over the incoming RAC, making it into a `FrameMaker` and then making the frame with that.  It's just a simple transform over the sequence.

Reply:
I agree that the code is trying to do things roundaboutly. I did try using `FrameMaker,` however, due to the `RowsAndColumnsUnravelingQueryRunner`, the sequence of RACs is a sequence of rows. Therefore it throws a `ClassCastException`:
```
class [Ljava.lang.Object; cannot be cast to class org.apache.druid.query.rowsandcols.RowsAndColumns
```

Setting `unravel` to false in the `ClientQuerySegmentWalker` when we know the results would be materialized as frames pose the same problem when trying to materialize the results as rows `resultsAsArrays`. The way I see it, the `ClientQuerySegmentWalker` should know if it's trying to materialize the results as arrays or frames, and depending on that set `unravel` to true/false. 

This leaks the implementation of the window tool chest to the calling class (`ClientQuerySegmentWalker`) where it should have been abstracted out. Do let me know if that is fine, or if there's a better resolution. 

FWIW here's the cleaned up `implementation` of the `resultsAsFrames` in case I misinterpreted your comment:

```java
  {
    return Optional.of(
        resultSequence.map(
            rac -> {
              FrameMaker frameMaker = FrameMaker.fromRAC(rac);
              return new FrameSignaturePair(
                  frameMaker.toColumnBasedFrame(),
                  frameMaker.computeSignature()
              );
            }
        )
    );
```

Reply:
The above is complicated by the fact that if we are not able to materialize the results as frames, we fall back and materialize them as arrays. So, if we have a "not-unraveled" result sequence because we were expecting the results as frames but had to fallback, then the `resultAsArrays` would also need to handle the case when it is "not-unraveled" or convert the `Sequence<RowsAndColumns>` back to `Sequence<Object[]>`, i.e. unravel it

Reply:
There's the same trouble with the suggested resolution: There are two ways that a subquery's results can be materialized: 
a. `resultsAsArrays`/List<Object[]> - For limiting the results by row number 
b. `resultsAsFrames`/List<Frame> - For limiting the results by memory size

(a) isn't required functionally, since Frames also contain the knowledge of the number of rows, however, at the time 2nd was added, we needed a way to:
1. Fallback if we encounter any difficulty while materializing the results as frames, allowing the user to transition
2. Allow the tool chests to transition gradually to `resultsAsFrames` - Since it is an extension point and not all tool chests would have implemented the functionality to materialize the results as frames.

`subquery: true` doesn't contain all the information, if it needs to materialize the results as arrays or frames. For the former, the code is fine as is, however, with the latter we don't need to unravel. 

I think one more mistake that the subquery materialization code made is overreliance on the generic type, which doesn't work here.

Reply:
> the context flag about num bytes versus num rows is what determines which thing does what, so there's a thing that already describes how to do the transition

I thought about this, however, we can also have a cluster-level config that determines the limit, so we should be looking at that as well in the window tool chest, which seems uncool that the window tool chest has to determine what to do.

>  If we just blindly try one, fail and then do the other, that will show up to users as a performance hit because they have no clue that there's this rando intermediate logic that is failing for a reason

Fallback is mostly for when the types aren't known. I agree that it is a performance hit, but at the time this feature was added, the signature informed by the tool chest didn't need to have a type. Scan queries only had knowledge of the column names (and not types), group by/time series... etc. toolchests could return `null` for the aggregator's dimensions. The fallback was present for these cases, where it's easy to detect the failure relatively early in the whole subquery processing flow. Fallback meant that transitioning from row -> byte based limit was simple. There's an undocumented parameter that treated these null types as JSON types, but that had logical flaws of its own iirc. 

Removing the fallback would make the change much easier and I have a lot more confidence that the query doesn't need to fallback (and we have the known cases before hand), however, I'd still like to keep it just in case for a while. I have an idea, and it depends on the fact that RACs can convert itself to frames properly, and window toolchests would never fall back. 

Thanks for the help!! I can work with the "serialization": "frame" parameter as a workaround to the current design choices.

Reply:
> I thought about this, however, we can also have a cluster-level config that determines the limit, so we should be looking at that as well in the window tool chest, which seems uncool that the window tool chest has to determine what to do.

If you have a context parameter, you automatically have a cluster-level config as well.  You can set the default context from cluster-level stuff.  If we are creating a context parameter, we shouldn't have yet-another random config to set things, that just makes it so much harder to figure out what is set how.

Additionally, there's nothing wrong with the window query tool chest being asked what to do, that's good.  The queries should be more involved in things because that allows them to optimize stuff based on what they know they need to do.  To wit, the whole reason you are having problems with this PR is because there was an attempt to take the decision *away* from the query.  If the code had been written in a way that gave the query control from the first place, you wouldn't have this problem.  If you are worried about "but then everyone has to implement the same thing", if it is truly the exact same implementation for every, single, call site, then it's incredibly easy to just have a static helper that all of the call sites can use.  Yes, each place has to call the helper, but the amount of time it takes to make a few places call a helper is significantly less than the amount of time it takes to work around an abstraction that isn't in the right place.

Reply:
I have added a context parameter (https://github.com/apache/druid/pull/16800/commits/ea1d57edf617e855bb35ca62d89519467debb797#diff-a31fa1872294019b29fc9940740cd66774a6af7430f8b7f3d4bc95ce3b5d6380R28) and I wanted a buy-in about its usefulness. 
Either
a. "serialization: frames/rows" 
or b. "subquery":true, along with `maxSubqueryBytes/maxSubqueryRows`
would be enough to disambiguate between the two ways in which the window toolchest can return the results.